File: libs/selenium-interface/mapgrab-selenium/src/main/java/com/mapgrab/selenium/MapController.java
package com.mapgrab.selenium;

import java.util.HashMap;
import java.util.Map;

import org.openqa.selenium.JavascriptExecutor;
import org.openqa.selenium.WebDriver;

import com.fasterxml.jackson.annotation.JsonInclude.Include;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.type.TypeFactory;
import com.mapgrab.selenium.controller.FitBoundsOptions;
import com.mapgrab.selenium.controller.LngLat;
import com.mapgrab.selenium.controller.MapBounds;
import com.mapgrab.selenium.controller.MutationState;
import com.mapgrab.selenium.controller.ScreenPoint;
import com.mapgrab.selenium.controller.SetAbsoluteViewOptions;
import com.mapgrab.selenium.controller.SetViewOptions;

public class MapController {
  protected final JavascriptExecutor executor;
  protected final String mapId;

  public MapController(WebDriver driver, String mapId) {
    this.mapId = mapId;
    this.executor = (JavascriptExecutor) driver;
  }

  public void waitToMapStable() {
    this.waitToMapInterface();

    this.executor.executeAsyncScript("""
          const mapId = arguments[0];
          const resolve = arguments[arguments.length - 1];

          const isStablePromise = __MAPGRAB__.getMapInterface(mapId).controller.waitToMapStable();
          isStablePromise && isStablePromise.then ? isStablePromise.then(resolve) : resolve();
        """,
        mapId);
  }

  public void enableInspector() {
    this.waitToMapInterface();

    this.executor.executeScript(
        """
               __MAPGRAB__.enableInspector();
            """);
  }

  public void disableInspector() {
    this.waitToMapInterface();

    this.executor.executeScript(
        """
               __MAPGRAB__.disableInspector();
            """);
  }

  public void waitToMapLoaded() {
    this.waitToMapInterface();

    this.executor.executeAsyncScript(
        """
              const mapId = arguments[0];
              const resolve = arguments[arguments.length - 1];

              const waitToMapLoadedPromise = __MAPGRAB__.getMapInterface(mapId).controller.waitToMapLoaded();
              waitToMapLoadedPromise && waitToMapLoadedPromise.then ? waitToMapLoadedPromise.then(resolve) : resolve()
            """,
        mapId);
  }

  public void setView(SetViewOptions options) {
    this.waitToMapLoaded();

    ObjectMapper mapper = new ObjectMapper();
    mapper.setSerializationInclusion(Include.NON_NULL);
    Map<String, Object> args = mapper.convertValue(options, Map.class);

    this.executor.executeScript(
        "console.log(arguments); __MAPGRAB__.getMapInterface(arguments[0]).controller.setView(arguments[1])",
        mapId, args);

    this.waitToMapStable();
  }

  public void setViewAbsolute(SetAbsoluteViewOptions options) {
    this.waitToMapLoaded();

    ObjectMapper mapper = new ObjectMapper();
    mapper.setSerializationInclusion(Include.NON_NULL);
    Map<String, Object> args = mapper.convertValue(options, Map.class);

    this.executor.executeScript(
        "console.log(arguments); __MAPGRAB__.getMapInterface(arguments[0]).controller.setViewAbsolute(arguments[1])",
        mapId, args);

    this.waitToMapStable();
  }

  public void fitMapToBounds(MapBounds bounds, FitBoundsOptions options) {
    ObjectMapper mapper = new ObjectMapper();
    mapper.setSerializationInclusion(Include.NON_NULL);
    Map<String, Object> optionsT = mapper.convertValue(options, Map.class);

    this.executor.executeScript(
        "__MAPGRAB__.getMapInterface(arguments[0]).controller.fitMapToBounds(arguments[1], arguments[2]) ",
        mapId, bounds.encodeToBoundsLike(), optionsT);

    this.waitToMapStable();
  }

  public void fitMapToBounds(MapBounds bounds) {
    this.fitMapToBounds(bounds, null);
  }

  public void fitMapToBoundingBox(BoundingBox bounds, FitBoundsOptions options) {
    ObjectMapper mapper = new ObjectMapper();
    mapper.setSerializationInclusion(Include.NON_NULL);
    Map<String, Object> boundsT = mapper.convertValue(bounds, Map.class);
    Map<String, Object> optionsT = mapper.convertValue(options, Map.class);

    this.executor.executeScript(
        "__MAPGRAB__.getMapInterface(arguments[0]).controller.fitMapToBoundingBox(arguments[1], arguments[2]) ",
        mapId, boundsT, optionsT);

    this.waitToMapStable();
  }

  public void fitMapToBoundingBox(BoundingBox bounds) {
    this.fitMapToBoundingBox(bounds, null);
  }

  public void setBackgroundColor(String color) {
    this.executor.executeScript(
        "__MAPGRAB__.getMapInterface(arguments[0]).controller.setBackgroundColor(arguments[1])",
        mapId, color);

    this.waitToMapStable();
  }

  public void removeBackground() {
    this.executor.executeScript(
        "__MAPGRAB__.getMapInterface(arguments[0]).controller.removeBackground()",
        mapId);

    this.waitToMapStable();
  }

  public ScreenPoint projectLngLatToScreenPoint(LngLat lngLat) {
    this.waitToMapStable();

    ObjectMapper mapper = new ObjectMapper();
    Map<String, Object> args = mapper.convertValue(lngLat, Map.class);

    Object res = this.executor.executeScript(
        "return __MAPGRAB__.getMapInterface(arguments[0]).controller.projectLngLatToScreenPoint(arguments[1])",
        mapId, args);

    return mapper.convertValue(res, ScreenPoint.class);
  }

  public LngLat unprojectScreenPointToLngLat(ScreenPoint point) {
    this.waitToMapStable();

    ObjectMapper mapper = new ObjectMapper();
    Map<String, Object> args = mapper.convertValue(point, Map.class);

    Object res = this.executor.executeScript(
        "return __MAPGRAB__.getMapInterface(arguments[0]).controller.unprojectScreenPointToLngLat(arguments[1])",
        mapId, args);

    return mapper.convertValue(res, LngLat.class);
  }

  public HashMap<String, MutationState> exposeLayers(String[] layersToExpose, String[] layersToHide) {
    this.waitToMapStable();

    Object res = this.executor.executeScript(
        "return __MAPGRAB__.getMapInterface(arguments[0]).controller.exposeLayers(arguments[1], arguments[2] || 'allOther')",
        mapId, layersToExpose, layersToHide);

    this.waitToMapStable();

    ObjectMapper mapper = new ObjectMapper();
    TypeFactory factory = TypeFactory.defaultInstance();

    return mapper.convertValue(res, factory.constructMapType(HashMap.class, String.class, MutationState.class));
  }

  public HashMap<String, MutationState> exposeLayers(String[] layersToExpose) {
    return this.exposeLayers(layersToExpose, null);
  }

  public void revertExposeLayers(HashMap<String, MutationState> exposeMutationState) {
    ObjectMapper mapper = new ObjectMapper();
    Map<String, Object> exposeMutationStateT = mapper.convertValue(exposeMutationState, Map.class);

    this.executor.executeScript(
        "__MAPGRAB__.getMapInterface(arguments[0]).controller.revertExposeLayers(arguments[1])",
        mapId, exposeMutationStateT);

    this.waitToMapStable();
  }

  private void waitToMapInterface() {
    this.executor.executeAsyncScript("""
        var callback = arguments[arguments.length - 1];

        if (window.__MAPGRAB__ && window.__MAPGRAB__.getMapInterface(arguments[0])) {
          callback();
          return;
        }

              const h = () => {
                if (window.__MAPGRAB__ && window.__MAPGRAB__.getMapInterface(arguments[0])) {
                   window.removeEventListener('__MAPGRAB__::MAP_INTERFACE_INIT', h);
                      callback();
                }
              }

              window.addEventListener('__MAPGRAB__::MAP_INTERFACE_INIT', h);
          """, mapId);
  }

}

File: libs/cypress-interface/src/lib/map-controller.ts
import { MapGrabEvents } from '@mapgrab/map-interface-types';

import type {
  MapControllerInterface as MapGrabControllerInterface,
  SetViewOptions,
  LngLatBoundsLike,
  FitBoundsOptions,
  ExposeMutationResult,
  MapType,
  MapRectInterface,
  MapGrabPublicInterfaceI,
  MapInterfaceI,
  LngLatInterface,
  PointLike,
  PointInterface,
} from '@mapgrab/map-interface-types';

export class MapController {
  constructor(
    protected readonly cy: Cypress.cy,
    protected readonly _window: Window,
    protected readonly mapId: string
  ) {}

  public setView(options: SetViewOptions) {
    return this.executeOnController((controller) => {
      controller.setView(options);
    });
  }

  public setViewAbsolute(options: SetViewOptions) {
    return this.executeOnController((controller) => {
      controller.setViewAbsolute(options);
    });
  }

  public fitMapToBounds(bounds: LngLatBoundsLike, options?: FitBoundsOptions) {
    return this.executeOnController((controller) => {
      controller.fitMapToBounds(bounds, options);
    });
  }

  public fitMapToBoundingBox(bbox: MapRectInterface, options?: FitBoundsOptions) {
    return this.executeOnController((controller) => {
      controller.fitMapToBoundingBox(bbox, options);
    });
  }

  public exposeLayers(
    layersToExpose: string[],
    layersToHide: string[] | 'allOther' = 'allOther'
  ): Cypress.Chainable<ExposeMutationResult> {
    return this.executeOnController((controller) => {
      return controller.exposeLayers(layersToExpose, layersToHide);
    });
  }

  public revertExposeLayers(exposeMutationState: ExposeMutationResult): Cypress.Chainable<void> {
    return this.executeOnController((controller) => {
      return controller.revertExposeLayers(exposeMutationState);
    });
  }

  public setBackgroundColor(backgroundColor: string): Cypress.Chainable<void> {
    return this.executeOnController((controller) => {
      return controller.setBackgroundColor(backgroundColor);
    });
  }

  public removeBackground(): Cypress.Chainable<void> {
    return this.executeOnController((controller) => {
      return controller.removeBackground();
    });
  }

  public projectLngLatToScreenPoint(lngLat: LngLatInterface): Cypress.Chainable<PointLike> {
    return this.executeOnController((controller) => {
      return controller.projectLngLatToScreenPoint(lngLat);
    });
  }

  public unprojectLngLatToScreenPoint(point: PointInterface): Cypress.Chainable<LngLatInterface> {
    return this.executeOnController((controller) => {
      return controller.unprojectScreenPointToLngLat(point);
    });
  }

  public waitToMapLoaded(opts?: { timeout?: number }) {
    return this.getMapController().then((c) => this.cy.wrap(c.waitToMapLoaded(), opts));
  }

  public waitToMapRepaint(opts?: { timeout?: number }) {
    return this.getMapController().then((c) => this.cy.wrap(c.waitToMapRepaint(), opts));
  }

  public waitToMapStable(opts?: { timeout?: number }) {
    return this.getMapController().then((map) => this.cy.wrap(map.waitToMapStable(), opts));
  }

  public getMapInstance(): Cypress.Chainable<MapType> {
    return this.getMapInterface().then((i) => i.map);
  }

  public getMapController(): Cypress.Chainable<MapGrabControllerInterface> {
    return this.getMapInterface().then((i) => i.controller);
  }

  public enableInspector() {
    return this.getWindow().then(() => {
      return this.getMapGrabHandle().then((mapGrab) => {
        mapGrab.enableInspector();
      });
    });
  }

  public disableInspector() {
    return this.getWindow().then(() => {
      return this.getMapGrabHandle().then((mapGrab) => {
        mapGrab.disableInspector();
      });
    });
  }

  public getMapInterface(): Cypress.Chainable<MapInterfaceI> {
    return this.getWindow().then((_window) => {
      return this.getMapGrabHandle().then((mapGrab): Cypress.Chainable<MapInterfaceI> => {
        const getInterface = () => mapGrab.getMapInterface(this.mapId),
          currentInterface = getInterface();

        if (currentInterface) {
          return this.cy.wrap(currentInterface, { log: false });
        }

        return this.cy.wrap<Promise<MapInterfaceI>, MapInterfaceI>(
          new Promise<MapInterfaceI>((resolve) => {
            const handler = () => {
              const currentInterface = getInterface();

              if (currentInterface) {
                _window.removeEventListener(MapGrabEvents.MAP_INTERFACE_INIT, handler);

                resolve(currentInterface);

                return;
              }
            };

            _window.addEventListener(MapGrabEvents.MAP_INTERFACE_INIT, handler);
            handler();
          }),
          { log: false }
        );
      });
    });
  }

  public getMapGrabHandle(): Cypress.Chainable<MapGrabPublicInterfaceI> {
    return this.getWindow().its('__MAPGRAB__', { log: false });
  }

  protected getWindow(): Cypress.Chainable<Window> {
    return this.cy.wrap(this._window, { log: false });
  }

  protected executeOnController(callback: (controller: MapGrabControllerInterface) => any | Cypress.Chainable<any>) {
    return this.waitToMapLoaded().then(() => {
      return this.getMapController().then((controller) => {
        const result = callback(controller);

        return this.cy
          .wrap(result, { log: false })
          .then(() => this.waitToMapStable())
          .then(() => result);
      });
    });
  }
}
